File: src/main/java/dev/kikugie/elytratrims/common/access/FeatureAccess.java
package dev.kikugie.elytratrims.common.access;

import com.mojang.datafixers.util.Pair;
import dev.kikugie.elytratrims.common.plugin.ModStatus;
import io.github.apfelrauber.stacked_trims.ArmorTrimList;
import net.minecraft.block.entity.BannerBlockEntity;
import net.minecraft.block.entity.BannerPattern;
import net.minecraft.item.BannerItem;
import net.minecraft.item.BlockItem;
import net.minecraft.item.ItemStack;
import net.minecraft.item.trim.ArmorTrim;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.nbt.NbtInt;
import net.minecraft.nbt.NbtList;
import net.minecraft.registry.DynamicRegistryManager;
import net.minecraft.registry.entry.RegistryEntry;
import net.minecraft.util.DyeColor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

/*? if >1.20.4 {*//*
import net.minecraft.component.DataComponentTypes;
import net.minecraft.component.type.BannerPatternsComponent;
import net.minecraft.component.type.DyedColorComponent;
import net.minecraft.component.type.NbtComponent;
*//*?} */

import static java.util.Collections.emptyList;

@SuppressWarnings("CommentedOutCode")
public class FeatureAccess {
    /*? if <=1.20.4*/
    private static final net.minecraft.item.DyeableItem DYEABLE = new net.minecraft.item.DyeableItem() {};

    public static List<BannerLayer> getPatterns(ItemStack stack) {
        /*? if <=1.20.4 {*/
        NbtList patterns = BannerBlockEntity.getPatternListNbt(stack);
        if (patterns == null) return emptyList();

        NbtCompound nbt = BlockItem.getBlockEntityNbt(stack);
        if (nbt == null) return emptyList();
        DyeColor baseColor = nbt.contains("Base") ? DyeColor.byId(nbt.getInt("Base")) : DyeColor.WHITE;
        List<BannerLayer> out = new ArrayList<>();
        for (Pair<RegistryEntry<BannerPattern>, DyeColor> entry : BannerBlockEntity.getPatternsFromNbt(baseColor, patterns))
            out.add(new BannerLayer(entry.getFirst(), entry.getSecond()));
        return out;
        /*?} else {*//*
        var patterns = stack.get(DataComponentTypes.BANNER_PATTERNS).layers();
        if (patterns == null || patterns.isEmpty()) return emptyList();

        List<BannerLayer> out = new ArrayList<>();
        for (BannerPatternsComponent.Layer layer : patterns)
            out.add(new BannerLayer(layer.pattern(), layer.color()));
        return out;
        *//*?} */
    }

    public static void setPatterns(ItemStack target, ItemStack source) {
        assert source.getItem() instanceof BannerItem;
        /*? if <=1.20.4 {*/
        NbtList patterns = BannerBlockEntity.getPatternListNbt(source);
        if (patterns != null) {
            DyeColor color = ((BannerItem) source.getItem()).getColor();
            NbtCompound storage = target.getOrCreateSubNbt("BlockEntityTag");
            storage.put("Patterns", patterns);
            storage.put("Base", NbtInt.of(color.getId()));
            setColor(target, color.getFireworkColor());
        } else if (target.getSubNbt("BlockEntityTag") != null && target.getSubNbt("BlockEntityTag").isEmpty()) {
            target.removeSubNbt("BlockEntityTag");
            setColor(target, 0);
        }
        /*?} else {*//*
        var patterns = source.getComponents().filtered(t -> t == DataComponentTypes.BANNER_PATTERNS);
        target.applyComponentsFrom(patterns);
        *//*?} */
    }

    public static int getColor(ItemStack stack) {
        /*? if <=1.20.4*/
        return DYEABLE.hasColor(stack) ? DYEABLE.getColor(stack) : 0;
        /*? if >1.20.4*/
        /*return DyedColorComponent.getColor(stack, 0);*/
    }

    public static void setColor(ItemStack target, int color) {
        /*? if <=1.20.4 {*/
        if (color != 0) DYEABLE.setColor(target, color);
        else DYEABLE.removeColor(target);
        /*?} else {*/
        /*target.set(DataComponentTypes.DYED_COLOR, new DyedColorComponent(color, true));*/
        /*?} */
    }

    public static List<ArmorTrim> getTrims(ItemStack stack, DynamicRegistryManager registry) {
        /*? if <=1.20.4 {*/
        if (ModStatus.isLoaded("stacked-armor-trims"))
            return ArmorTrimList.getTrims(registry, stack).orElse(emptyList());
        Optional<ArmorTrim> trim = ArmorTrim.getTrim(registry, stack);
        return trim.map(Collections::singletonList).orElse(emptyList());
        /*?} else {*//*
        var trim = stack.get(DataComponentTypes.TRIM);
        return trim != null ? Collections.singletonList(trim) : emptyList();
        *//*?} */
    }

    public static boolean hasGlow(ItemStack stack) {
        /*? if <=1.20.4 {*/
        NbtCompound nbtCompound = stack.getSubNbt("display");
        return nbtCompound != null && nbtCompound.contains("glow", 1) && nbtCompound.getBoolean("glow");
        /*?} else {*//*
        var data = stack.get(DataComponentTypes.CUSTOM_DATA);
        return data != null && data.getNbt().getBoolean("glow");
        *//*?} */
    }

    public static void setGlow(ItemStack stack, boolean glow) {
        /*? if <=1.20.4 {*/
        if (glow) stack.getOrCreateSubNbt("display").putBoolean("glow", true);
        else {
            NbtCompound nbtCompound = stack.getSubNbt("display");
            if (nbtCompound != null && nbtCompound.contains("color")) {
                nbtCompound.remove("glow");
            }
        }
        /*?} else {*//*
        var data = stack.get(DataComponentTypes.CUSTOM_DATA);
        if (data == null && glow) {
            data = NbtComponent.DEFAULT;
            stack.set(DataComponentTypes.CUSTOM_DATA, data);
        }
        if (glow) data.getNbt().putBoolean("glow", true);
        else data.getNbt().remove("glow");
        *//*?} */
    }
}

File: src/main/kotlin/dev/kikugie/elytratrims/common/access/FeatureAccess.kt
package dev.kikugie.elytratrims.common.access

import dev.kikugie.elytratrims.common.util.toARGB
import dev.kikugie.elytratrims.platform.ModStatus
import io.github.apfelrauber.stacked_trims.ArmorTrimList
import net.minecraft.item.ItemStack
import net.minecraft.item.trim.ArmorTrim
import net.minecraft.registry.DynamicRegistryManager

/*? if <=1.20.4 {*/
import net.minecraft.block.entity.BannerBlockEntity
import net.minecraft.item.BannerItem
import net.minecraft.item.BlockItem
import net.minecraft.nbt.NbtCompound
import net.minecraft.util.DyeColor

object FeatureAccess : IFeatureAccess {
    private val DYEABLE = object : net.minecraft.item.DyeableItem {}
    override fun ItemStack.getTrims(manager: DynamicRegistryManager): List<ArmorTrim> =
        getArmorTrimList(this, manager) ?:
        getTrim(manager).orElse(null)?.let(::listOf) ?:
        emptyList()

    override fun ItemStack.getPatterns(): List<BannerLayer> {
        val nbt = BannerBlockEntity.getPatternListNbt(this) ?: return emptyList()
        return BannerBlockEntity.getPatternsFromNbt(DyeColor.WHITE, nbt).drop(1).map {
            BannerLayer(it.first, it.second)
        }
    }

    override fun ItemStack.getBaseColor(): Int = (item as? BannerItem)?.color?.fireworkColor ?: 0

    override fun ItemStack.setPatterns(source: ItemStack) {
        val nbt = BannerBlockEntity.getPatternListNbt(source) ?: return
        val container = BlockItem.getBlockEntityNbt(this) ?: NbtCompound()
        if (container.isEmpty) setSubNbt("BlockEntityTag", container)
        container.put("Patterns", nbt.copy())
    }

    override fun ItemStack.removePatterns() {
        val container = BlockItem.getBlockEntityNbt(this) ?: return
        container.remove("Patterns")
        if (container.isEmpty) removeSubNbt("BlockEntityTag")
    }

    override fun ItemStack.getColor() = (item as? BannerItem)?.color?.colorComponents?.toARGB() ?: 0

    override fun ItemStack.setColor(color: Int) {
        DYEABLE.setColor(this, color)
    }

    override fun ItemStack.removeColor() {
        DYEABLE.removeColor(this)
    }

    override fun ItemStack.hasGlow() = getSubNbt("display")?.getBoolean("glow") ?: false

    override fun ItemStack.addGlow() {
        getOrCreateSubNbt("display").putBoolean("glow", true)
    }

    override fun ItemStack.removeGlow() {
        val nbt = getSubNbt("display") ?: return
        nbt.remove("glow")
        if (nbt.isEmpty) removeSubNbt("display")
    }
}
/*?} else {*//*
import net.minecraft.component.DataComponentTypes
import net.minecraft.component.type.DyedColorComponent
import net.minecraft.component.type.NbtComponent
import net.minecraft.item.BannerItem

object FeatureAccess : IFeatureAccess {
    override fun ItemStack.getTrims(manager: DynamicRegistryManager): List<ArmorTrim> =
        if (ModStatus.isLoaded("stacked-armor-trims")) ArmorTrimList.getTrims(manager, this).orElse(emptyList())
        else get(DataComponentTypes.TRIM)?.let(::listOf) ?: emptyList()

    override fun ItemStack.getPatterns() = get(DataComponentTypes.BANNER_PATTERNS)?.layers?.map {
        BannerLayer(it.pattern, it.color)
    } ?: emptyList()

    override fun ItemStack.getBaseColor(): Int = (item as? BannerItem)?.color?.colorComponents?.toARGB() ?: 0

    override fun ItemStack.setPatterns(source: ItemStack) {
        applyComponentsFrom(source.components.filtered { it == DataComponentTypes.BANNER_PATTERNS })
    }

    override fun ItemStack.removePatterns() {
        remove(DataComponentTypes.BANNER_PATTERNS)
    }

    override fun ItemStack.getColor(): Int = DyedColorComponent.getColor(this, 0)

    override fun ItemStack.setColor(color: Int) {
        set(DataComponentTypes.DYED_COLOR, DyedColorComponent(color, true))
    }

    override fun ItemStack.removeColor() {
        remove(DataComponentTypes.DYED_COLOR)
    }

    // Why copy nbt if we don't modify it
    override fun ItemStack.hasGlow(): Boolean = get(DataComponentTypes.CUSTOM_DATA)?.nbt?.getBoolean("glow") ?: false

    override fun ItemStack.addGlow() {
        val data = (get(DataComponentTypes.CUSTOM_DATA) ?: NbtComponent.DEFAULT).copyNbt()
        data.putBoolean("glow", true)
        NbtComponent.set(DataComponentTypes.CUSTOM_DATA, this, data)
    }

    override fun ItemStack.removeGlow() {
        val data = get(DataComponentTypes.CUSTOM_DATA)?.copyNbt() ?: return
        data.remove("glow")
        NbtComponent.set(DataComponentTypes.CUSTOM_DATA, this, data)
    }
}
*//*?} */

private fun getArmorTrimList(stack: ItemStack, manager: DynamicRegistryManager): List<ArmorTrim>? =
    /*? if fabric {*/
    if (ModStatus.isLoaded("stacked-armor-trims")) ArmorTrimList.getTrims(manager, stack).orElse(null) else null
    /*?} else {*//*
    null
    *//*?} */

/*? if <=1.20.4 >=1.20.2 */
/*private fun ItemStack.getTrim(manager: DynamicRegistryManager) = ArmorTrim.getTrim(manager, this, true)*/
/*? if <1.20.2 */
private fun ItemStack.getTrim(manager: DynamicRegistryManager) = ArmorTrim.getTrim(manager, this)

private interface IFeatureAccess {
    fun ItemStack.getTrims(manager: DynamicRegistryManager): List<ArmorTrim>

    fun ItemStack.getPatterns(): List<BannerLayer>
    fun ItemStack.getBaseColor(): Int
    fun ItemStack.setPatterns(source: ItemStack)
    fun ItemStack.removePatterns()

    fun ItemStack.getColor(): Int
    fun ItemStack.setColor(color: Int)
    fun ItemStack.removeColor()

    fun ItemStack.hasGlow(): Boolean
    fun ItemStack.addGlow()
    fun ItemStack.removeGlow()
}